### ISSUE TITLE
hledger-ui --change, date: broken in master

### ISSUE THREAD

**Issue by simonmichael** [simonmichael/hledger#1387]
Two more issues I've just noticed, since hledger-ui-1.19.1:

- a date: query argument is having no effect, eg 
  `hledger-ui -f examples/sample.journal --register checking date:2008-01` 
  should show only one transaction
- `--change` is having no effect

Possibly more fallout from #1340 / related to #1383 @Xitian9 ?

It's unfortunate that we don't have tests for hledger-ui. I've just committed an old attempt, in case anyone feels like a challenge: https://github.com/simonmichael/hledger/tree/master/tests/ui

**Comment by Xitian9** [comment]
Thanks for pointing these out, I'll have a look.

I don't use hledger-ui myself, so though I try to do some basic testing on my own I'm not too familiar with what functionality to look for. Even if automatic testing is difficult or infeasible, perhaps a checklist along the lines of ‘perform the following actions and make sure everything works’ might be worthwhile. It could be run manually for any major updates, and before every release.

**Comment by simonmichael** [comment]
Good idea - I will make some human test scripts.

**Comment by simonmichael** [comment]
I have been looking at this (and I see vs code assigned it to me). I see the bug, though not yet the fix. At startup, hledger-ui does some shenanigans with the initial query generated from command line options and arguments: any date or depth terms are removed from the query and stored elsewhere (ReportOpts' period_ and depth_ fields), because there is special UI for date and depth (the number/+/- keys and the shift-up/down/left/right keys), and it's cleaner not to have redundant/conflicting date: and depth: terms in the query that users edit with the / key. This worked ok when the three screens regenerated their data query with queryFromOpts, because that pulled in the date and depth terms again. But since "ui,web: Adopt new ReportOptions interface, store literal query string in UIOpts." (fcbe511d) they are using the same filtered user query as their data query, with date and depth terms missing. I guess we fixed the depth issue in a different way since then but date is still a problem. This also causes the --change issue.

**Comment by simonmichael** [comment]
I guess this is why we had to add error calls in 74ce7be5 (#1383), because when adjusting date or depth in the UI it has to update the query, which currently means reparsing the whole thing from flags and query arguments.

That's the right thing to do when the user edits query terms with `/` - a parse error there can be reported and rejected. But when they adjust depth and date via the UI, that's a safe change and we shouldn't have to reparse. I think in this case the screens probably can adjust the existing query rather than regenerating it. (And perhaps we can revert 74ce7be5.)

**Comment by simonmichael** [comment]
I guess the proof will be in the pudding. I'll try again another day, or feel free to take a crack at it.

**Comment by simonmichael** [comment]
9566a1a in the simonmichael/issue1387 branch is a possible quick fix. More cleanup needed I think.

**Comment by simonmichael** [comment]
The fix seems ok so far. To save time I'm just going to merge it, and not spend more time on cleanup right now but leave these notes instead:

Where we went wrong, more succinctly: hledger-ui used to regenerate the query, when needed, from all arguments and options (queryFromOpts). We made it use a more static query, which is not yet properly updated by all UI actions.

Some notes and terminology for queries:

The *startup query* is the Query generated at program startup, from
command line options, arguments, and the current date. hledger CLI
uses this.

hledger-ui/hledger-web allow the query to be changed at will, creating
a new *runtime query* each time.

The startup query or part of it can be used as a *constraint query*, 
limiting all runtime queries. hledger-web does this with the startup 
report period, never showing transactions outside those dates.
hledger-ui does not do this.

A query is a combination of multiple subqueries/terms, which are
generated from command line options and arguments, ui/web app runtime
state, and/or the current date.

Some subqueries are generated by parsing freeform user input, which
can fail. We don't want hledger users to see such failures except:

1. at program startup, in which case the program exits
2. after entering a new freeform query in hledger-ui/web, in which case the query change is rejected

So we should parse those kinds of subquery only at those times. Any
subqueries which do not require parsing can be kept separate. And
these can be combined to make the full query when needed, eg when
hledger-ui screens are generating their data.
